File: ts/src/registry.ts
/** Module-type registry (plugin surface). */

import { ValidationError } from "./errors.js";
import type {
  Adaptation,
  Module,
  ModuleType,
  Priority,
  SpeechMode,
} from "./models.js";

export type Frontmatter = Record<string, unknown>;

export type Validator = (fm: Frontmatter, body: string) => string[];

export interface TypeSpec {
  type: ModuleType;
  slot: string;
  validateFrontmatter: Validator;
  maxCount?: number;
}

function requireName(fm: Frontmatter): string[] {
  if (!fm.name) return ["missing required field: name"];
  return [];
}

function validateIdentity(fm: Frontmatter, _body: string): string[] {
  return requireName(fm);
}

function validateRole(fm: Frontmatter, _body: string): string[] {
  const errors = requireName(fm);
  if (fm.tools != null && !Array.isArray(fm.tools)) {
    errors.push("role.tools must be a list");
  }
  return errors;
}

function validateTrait(fm: Frontmatter, _body: string): string[] {
  const errors = requireName(fm);
  const priority = fm.priority;
  if (priority == null) {
    errors.push("trait requires priority (high|medium|low)");
  } else if (!["high", "medium", "low"].includes(String(priority))) {
    errors.push(`invalid trait.priority: ${JSON.stringify(priority)}`);
  }
  const conflicts = fm.conflicts ?? [];
  if (!Array.isArray(conflicts)) {
    errors.push("trait.conflicts must be a list");
  }
  return errors;
}

function validateImportFields(fm: Frontmatter, body: string): string[] {
  const errors: string[] = [];
  const source = fm.source;
  const adaptation = fm.adaptation;
  if (
    adaptation != null &&
    !["as-is", "extracted"].includes(String(adaptation))
  ) {
    errors.push(`invalid adaptation: ${JSON.stringify(adaptation)}`);
  }
  if (adaptation === "as-is") {
    if (!source) errors.push("adaptation as-is requires source");
    if (body.trim()) errors.push("adaptation as-is requires an empty overlay body");
  }
  if (source && adaptation == null) {
    errors.push("source requires adaptation (as-is|extracted)");
  }
  return errors;
}

function withImport(base: Validator): Validator {
  return (fm, body) => [...base(fm, body), ...validateImportFields(fm, body)];
}

function validateSpeech(fm: Frontmatter, body: string): string[] {
  const errors = requireName(fm);
  const mode = fm.mode ?? "prompt";
  if (!["prompt", "rewriter"].includes(String(mode))) {
    errors.push(`invalid speech.mode: ${JSON.stringify(mode)}`);
  }
  errors.push(...validateImportFields(fm, body));
  return errors;
}

function validateRelationship(fm: Frontmatter, _body: string): string[] {
  const errors = requireName(fm);
  if (!fm.agent) errors.push("relationship requires agent");
  if (!fm.status) errors.push("relationship requires status");
  return errors;
}

function validateOutputRules(fm: Frontmatter, _body: string): string[] {
  return requireName(fm);
}

export const BUILTIN_TYPES: Record<string, TypeSpec> = {
  identity: {
    type: "identity",
    slot: "identity",
    validateFrontmatter: withImport(validateIdentity),
    maxCount: 1,
  },
  role: {
    type: "role",
    slot: "role",
    validateFrontmatter: withImport(validateRole),
    maxCount: 1,
  },
  trait: {
    type: "trait",
    slot: "traits",
    validateFrontmatter: withImport(validateTrait),
  },
  speech: {
    type: "speech",
    slot: "speech",
    validateFrontmatter: validateSpeech,
  },
  relationship: {
    type: "relationship",
    slot: "relationships",
    validateFrontmatter: withImport(validateRelationship),
  },
  output_rules: {
    type: "output_rules",
    slot: "output_rules",
    validateFrontmatter: withImport(validateOutputRules),
    maxCount: 1,
  },
};

export class TypeRegistry {
  private specs: Map<string, TypeSpec>;

  constructor(specs: Record<string, TypeSpec> = BUILTIN_TYPES) {
    this.specs = new Map(Object.entries(specs));
  }

  register(spec: TypeSpec): void {
    this.specs.set(spec.type, spec);
  }

  get(typeName: string): TypeSpec | undefined {
    return this.specs.get(typeName);
  }

  require(typeName: string): TypeSpec {
    const spec = this.get(typeName);
    if (!spec) throw new ValidationError(`unknown type: ${JSON.stringify(typeName)}`);
    return spec;
  }

  knownTypes(): Set<string> {
    return new Set(this.specs.keys());
  }
}

export const DEFAULT_REGISTRY = new TypeRegistry();

export function applyFrontmatter(module: Module, fm: Frontmatter): void {
  if (module.type === "trait") {
    module.priority = String(fm.priority) as Priority;
    module.conflicts = Array.isArray(fm.conflicts)
      ? fm.conflicts.map(String)
      : [];
  } else if (module.type === "role") {
    module.tools = Array.isArray(fm.tools) ? fm.tools.map(String) : [];
  } else if (module.type === "speech") {
    module.mode = String(fm.mode ?? "prompt") as SpeechMode;
  } else if (module.type === "relationship") {
    module.agent = String(fm.agent);
    module.status = String(fm.status);
  }

  if (fm.source) module.source = String(fm.source);
  if (fm.adaptation) module.adaptation = String(fm.adaptation) as Adaptation;
  if (fm.origin) module.origin = String(fm.origin);
}

File: src/persona_composer/registry.py
"""Module-type registry (plugin surface)."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass

from persona_composer.errors import ValidationError
from persona_composer.models import Adaptation, Module, ModuleType, Priority, SpeechMode

Frontmatter = dict

Validator = Callable[[Frontmatter, str], list[str]]


@dataclass(frozen=True)
class TypeSpec:
    type: ModuleType
    slot: str
    validate_frontmatter: Validator
    max_count: int | None = None  # None = unbounded; 1 = at most one


def _require_name(fm: Frontmatter) -> list[str]:
    if not fm.get("name"):
        return ["missing required field: name"]
    return []


def _validate_identity(fm: Frontmatter, _body: str) -> list[str]:
    return _require_name(fm)


def _validate_role(fm: Frontmatter, _body: str) -> list[str]:
    errors = _require_name(fm)
    tools = fm.get("tools")
    if tools is not None and not isinstance(tools, list):
        errors.append("role.tools must be a list")
    return errors


def _validate_trait(fm: Frontmatter, _body: str) -> list[str]:
    errors = _require_name(fm)
    priority = fm.get("priority")
    if priority is None:
        errors.append("trait requires priority (high|medium|low)")
    elif priority not in {p.value for p in Priority}:
        errors.append(f"invalid trait.priority: {priority!r}")
    conflicts = fm.get("conflicts", [])
    if conflicts is None:
        conflicts = []
    if not isinstance(conflicts, list):
        errors.append("trait.conflicts must be a list")
    return errors


def _validate_speech(fm: Frontmatter, body: str) -> list[str]:
    errors = _require_name(fm)
    mode = fm.get("mode", SpeechMode.PROMPT.value)
    if mode not in {m.value for m in SpeechMode}:
        errors.append(f"invalid speech.mode: {mode!r}")
    errors.extend(_validate_import_fields(fm, body))
    return errors


def _validate_relationship(fm: Frontmatter, _body: str) -> list[str]:
    errors = _require_name(fm)
    if not fm.get("agent"):
        errors.append("relationship requires agent")
    if not fm.get("status"):
        errors.append("relationship requires status")
    return errors


def _validate_output_rules(fm: Frontmatter, _body: str) -> list[str]:
    return _require_name(fm)


def _validate_import_fields(fm: Frontmatter, body: str) -> list[str]:
    errors: list[str] = []
    source = fm.get("source")
    adaptation = fm.get("adaptation")
    if adaptation is not None and adaptation not in {a.value for a in Adaptation}:
        errors.append(f"invalid adaptation: {adaptation!r}")
    if adaptation == Adaptation.AS_IS.value:
        if not source:
            errors.append("adaptation as-is requires source")
        if body.strip():
            errors.append("adaptation as-is requires an empty overlay body")
    if source and adaptation is None:
        errors.append("source requires adaptation (as-is|extracted)")
    return errors


def _validate_with_import(
    base: Validator,
) -> Validator:
    def _wrapped(fm: Frontmatter, body: str) -> list[str]:
        return base(fm, body) + _validate_import_fields(fm, body)

    return _wrapped


BUILTIN_TYPES: dict[str, TypeSpec] = {
    ModuleType.IDENTITY.value: TypeSpec(
        type=ModuleType.IDENTITY,
        slot="identity",
        validate_frontmatter=_validate_with_import(_validate_identity),
        max_count=1,
    ),
    ModuleType.ROLE.value: TypeSpec(
        type=ModuleType.ROLE,
        slot="role",
        validate_frontmatter=_validate_with_import(_validate_role),
        max_count=1,
    ),
    ModuleType.TRAIT.value: TypeSpec(
        type=ModuleType.TRAIT,
        slot="traits",
        validate_frontmatter=_validate_with_import(_validate_trait),
    ),
    ModuleType.SPEECH.value: TypeSpec(
        type=ModuleType.SPEECH,
        slot="speech",
        validate_frontmatter=_validate_speech,
    ),
    ModuleType.RELATIONSHIP.value: TypeSpec(
        type=ModuleType.RELATIONSHIP,
        slot="relationships",
        validate_frontmatter=_validate_with_import(_validate_relationship),
    ),
    ModuleType.OUTPUT_RULES.value: TypeSpec(
        type=ModuleType.OUTPUT_RULES,
        slot="output_rules",
        validate_frontmatter=_validate_with_import(_validate_output_rules),
        max_count=1,
    ),
}


class TypeRegistry:
    """Open registry of module types."""

    def __init__(self, specs: dict[str, TypeSpec] | None = None) -> None:
        self._specs = dict(specs or BUILTIN_TYPES)

    def register(self, spec: TypeSpec) -> None:
        self._specs[spec.type.value] = spec

    def get(self, type_name: str) -> TypeSpec | None:
        return self._specs.get(type_name)

    def require(self, type_name: str) -> TypeSpec:
        spec = self.get(type_name)
        if spec is None:
            raise ValidationError(f"unknown type: {type_name!r}")
        return spec

    def known_types(self) -> set[str]:
        return set(self._specs)


DEFAULT_REGISTRY = TypeRegistry()


def apply_frontmatter(module: Module, fm: Frontmatter) -> None:
    """Fill type-specific fields on a Module from frontmatter (post-parse)."""
    if module.type == ModuleType.TRAIT:
        module.priority = Priority(fm["priority"])
        conflicts = fm.get("conflicts") or []
        module.conflicts = list(conflicts)
    elif module.type == ModuleType.ROLE:
        module.tools = list(fm.get("tools") or [])
    elif module.type == ModuleType.SPEECH:
        module.mode = SpeechMode(fm.get("mode", SpeechMode.PROMPT.value))
    elif module.type == ModuleType.RELATIONSHIP:
        module.agent = str(fm["agent"])
        module.status = str(fm["status"])

    if fm.get("source"):
        module.source = str(fm["source"])
    if fm.get("adaptation"):
        module.adaptation = Adaptation(fm["adaptation"])
    if fm.get("origin"):
        module.origin = str(fm["origin"])
